### PR TITLE
Bandwidth check for gRPC copies

### PR DESCRIPTION
**Pull request by JuanCarlosGI**
The logic for the bandwidth check was taken from: http://index/?leftProject=CloudBuild.Services.Common&leftSymbol=acmnpfersqik&file=Communication%5CFsServer%5CFsServerClient.cs (BandwidthCheckedCopyAsync)

This will cause a breaking change in CloudBuild, since MinimumSpeedInMbPerSec is now double? instead of double. The actual fix for the breaking change should be trivial.

Note: as part of the PR, I moved the gRPC copy client to ContentStore.Distributed, since it is only used there.

### CODE REVIEW

**Comment by ljcollins25** [comment]
> // Copyright (c) Microsoft. All rights reserved.

Is there a reason we can't move this to ContentStore rather than moving Grpc stuff up to Distributed layer?

---
Refers to: Public/Src/Cache/ContentStore/Distributed/Utilities/BandwidthChecker.cs:1 in 16d976d. [](commit_id = 16d976de9fb63270102d491d9ea17127b3311552, deletion_comment = False)

**Comment by JuanCarlosGI** [comment]
> // Copyright (c) Microsoft. All rights reserved.

There is no reason, other than I think Distributed is the logical place for this to live in.

---
In reply to: [533670847](https://github.com/microsoft/BuildXL/pull/898#issuecomment-533670847) [](ancestors = 533670847)

---
Refers to: Public/Src/Cache/ContentStore/Distributed/Utilities/BandwidthChecker.cs:1 in 16d976d. [](commit_id = 16d976de9fb63270102d491d9ea17127b3311552, deletion_comment = False)

File: Public/Src/Cache/ContentStore/Library/Utils/BandwidthChecker.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                var bytesCopied = endPosition - startPosition;
+                var speed = bytesCopied / timer.Elapsed.TotalSeconds / (1024 * 1024);
+                _historicalBandwidthLimitSource.AddBandwidthRecord(speed);
+            }
+            else
+            {
+                await Impl();
+            }
+
+            async Task Impl()
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Library/Utils/BandwidthChecker.cs]: nit: local functions should be lowercased

File: Public/Src/Cache/ContentStore/Library/Utils/IBandwidthLimitSource.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            _maxRecordsStored = maxRecordsStored;
+            _bandwidthRecords = new double[maxRecordsStored];
+        }
+
+        /// <summary>
+        /// Adds a record for future use.
+        /// </summary>
+        public void AddBandwidthRecord(double value)
+        {
+            // Do a quick check to protect ourselves against invalid data: NaN's, negative, and zero values.
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Library/Utils/IBandwidthLimitSource.cs]: double.NaN <= 0.0 is false.

So you're not protected here against NaN's.

File: Public/Src/Cache/ContentStore/Library/Utils/IBandwidthLimitSource.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        /// <inheritdoc />
+        public double GetMinimumSpeedInMbPerSec()
+        {
+            var statistics = GetBandwidthRecordStatistics();
+            if (statistics.Count < _maxRecordsStored / 2)
+            {
+                // If not enough records, don't impose a limit.
+                return (0.0);
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Library/Utils/IBandwidthLimitSource.cs]: nit: do you need braces here (and at line 78)

**Pull request merged**